util/db/cluster.go

```diff
@@ -106,68 +122,69 @@ type ClusterEvent struct {
-	listOpts := metav1.ListOptions{}
-	labelSelector := labels.NewSelector()
-	req, err := labels.NewRequirement(common.LabelKeySecretType, selection.Equals, []string{common.LabelValueSecretTypeCluster})
-	if err != nil {
-		return err
-	}
-	labelSelector = labelSelector.Add(*req)
-	listOpts.LabelSelector = labelSelector.String()
-	w, err := db.kubeclientset.CoreV1().Secrets(db.ns).Watch(listOpts)
-	if err != nil {
-		return err
-	}
-
+func (db *db) WatchClusters(ctx context.Context,
+	handleAddEvent func(cluster *appv1.Cluster),
+	handleModEvent func(oldCluster *appv1.Cluster, newCluster *appv1.Cluster),
+	handleDeleteEvent func(clusterServer string)) error {
 	localCls, err := db.GetCluster(ctx, common.KubernetesInternalAPIServerAddr)
 	if err != nil {
 		return err
 	}
+	handleAddEvent(localCls)
 
-	defer w.Stop()
-	watchErr := make(chan error)
-
-	// trigger callback with event for local cluster since it always considered added
-	callback(&ClusterEvent{Type: watch.Added, Cluster: localCls})
-
-	go func() {
-		for next := range w.ResultChan() {
-			secret, ok := next.Object.(*apiv1.Secret)
-			if !ok {
-				watchErr <- fmt.Errorf("failed to convert event object to secret")
-				return
-			}
-			cluster := secretToCluster(secret)
-
-			// change local cluster event to modified or deleted, since it cannot be re-added or deleted
-			if cluster.Server == common.KubernetesInternalAPIServerAddr {
-				if next.Type == watch.Deleted {
-					next.Type = watch.Modified
-					cluster = &localCluster
-				} else if next.Type == watch.Added {
-					if !reflect.DeepEqual(localCls.Config, cluster.Config) {
-						localCls = cluster
-						next.Type = watch.Modified
-					} else {
-						continue
+	clusterSecretListOptions := func(options *metav1.ListOptions) {
+		clusterLabelSelector := fields.ParseSelectorOrDie(common.LabelKeySecretType + "=" + common.LabelValueSecretTypeCluster)
+		options.LabelSelector = clusterLabelSelector.String()
+	}
+	clusterEventHandler := cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			if secretObj, ok := obj.(*apiv1.Secret); ok {
+				cluster := secretToCluster(secretObj)
+				if cluster.Server == common.KubernetesInternalAPIServerAddr {
+					if reflect.DeepEqual(localCls.Config, cluster.Config) {
+						return
 					}
-				} else {
 					localCls = cluster
+					// change local cluster event to modified or deleted, since it cannot be re-added or deleted
+					handleModEvent(localCls, cluster)
+					return
 				}
+				handleAddEvent(cluster)
 			}
-
-			callback(&ClusterEvent{
-				Type:    next.Type,
-				Cluster: cluster,
-			})
-		}
-		watchErr <- errors.New("secret watch closed")
+		},
+		DeleteFunc: func(obj interface{}) {
+			if secretObj, ok := obj.(*apiv1.Secret); ok {
+				if string(secretObj.Data["server"]) == common.KubernetesInternalAPIServerAddr {
+					// change local cluster event to modified or deleted, since it cannot be re-added or deleted
+					handleModEvent(nil, localCls)
```

Comment:
handleModEvent should not use nil for previous of updated cluster .

Reply:
In this scenario that deleted cluster is common.KubernetesInternalAPIServerAddr, I need to force to invalidate cache in the handleModEvent. I used nil as input of previous cluster to force it invalidate.

Reply:
```
DeleteFunc: func(obj interface{}) {
    if secretObj, ok := obj.(*apiv1.Secret); ok {
        if string(secretObj.Data["server"]) == common.KubernetesInternalAPIServerAddr {
            // change local cluster event to modified or deleted, since it cannot be re-added or deleted
            
            handleModEvent(localCls, localCluster)
            localCls = localCluster
        } else {
            handleDeleteEvent(string(secretObj.Data["server"]))
        }
    }
},
```